### PR TITLE
Port doc comment changes from docs repo

### PR DESCRIPTION
**Pull request by ghogen**
Fixes https://github.com/dotnet/msbuild/issues/7151

### Context
Doc comments are changed from time to time in the dotnet-api-docs repo. These changes need to be ported back to the MSBuild source, now that the MSBuild source is being used as the source-of-truth for MSBuild reference docs.

### Changes Made
Comment-only changes in a number of source files.

### Notes
The changes that were required can also be found here: https://github.com/dotnet/dotnet-api-docs/pull/7484 - this docs PR was necessary to prevent the dotnet-api-docs changes from being overwritten when the comments were re-read from MSBuild sources for the 17.0 version of MSBuild.

### CODE REVIEW

File: src/Framework/Sdk/SdkResolverContext.cs [hunk 1]
```diff
@@ -39,8 +39,11 @@ public abstract class SdkResolverContext
         ///     Version of MSBuild currently running.
         /// </summary>
         /// <remarks>
-        ///     File version based on commit height from our public git repository. This is informational
-        ///     and not equal to the assembly version.
+        ///    <format type="text/markdown"><![CDATA[
+        /// ## Remarks
+        ///     
+        /// File version is based on commit height from our public git repository. This is informational and not equal to the assembly version.
```
Comment (rainersigwald) [inline src/Framework/Sdk/SdkResolverContext.cs:45]:
```suggestion
        /// File version is informational and not equal to the assembly version.
```

(we don't use the commit height any more)

File: src/Tasks/CodeTaskFactory.cs [hunk 2]
```diff
@@ -170,7 +170,7 @@ public TaskPropertyInfo[] GetTaskParameters()
         }
 
         /// <summary>
-        /// Initialze the task factory
+        /// Initialzes the task factory.
```
Comment (rainersigwald) [inline src/Tasks/CodeTaskFactory.cs:173]:
```suggestion
        /// Initializes the task factory.
```

File: src/Tasks/Copy.cs [hunk 3]
```diff
@@ -101,26 +101,26 @@ public Copy()
...
         /// <summary>
-        /// Delay between any necessary retries.
+        /// Gets or sets the delay, in milliseconds, between any necessary retries.
         /// Defaults to <see cref="RetryDelayMillisecondsDefault">RetryDelayMillisecondsDefault</see>
         /// </summary>
         public int RetryDelayMilliseconds { get; set; }
 
         /// <summary>
-        /// Create Hard Links for the copied files rather than copy the files if possible to do so
+        /// Gets or sets a value that indicates whether to use Hard Links for the copied files
```
Comment (rainersigwald) [inline src/Tasks/Copy.cs:116]:
```suggestion
        /// Gets or sets a value that indicates whether to use hard links for the copied files
```

(this appears to be the Microsoft style per https://docs.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions)

File: src/Tasks/Copy.cs [hunk 4]
```diff
@@ -101,26 +101,26 @@ public Copy()
...
+        /// rather than copy the files, if it's possible to do so.
         /// </summary>
         public bool UseHardlinksIfPossible { get; set; }
 
         /// <summary>
-        /// Create Symbolic Links for the copied files rather than copy the files if possible to do so
+        /// Gets or sets a value that indicates whether to create Symbolic Links for the copied files
```
Comment (rainersigwald) [inline src/Tasks/Copy.cs:122]:
```suggestion
        /// Gets or sets a value that indicates whether to create symbolic links for the copied files
```

**Review by rainersigwald — APPROVED** [review]

**Comment by danmoseley** [comment]
@ghogen just curious, do you do this manually or with a tool like https://github.com/carlossanlop/DocsPortingTool ?

**Comment by ghogen** [comment]
@danmoseley I did this manually, but yes the DocsPortingTool is designed to handle this kind of thing. It looked like a bit too much overhead for such a small number of changes.

**Comment by danmoseley** [comment]
@ghogen that's cool, just checking we weren't missing an opportunity to share tech.

File: src/Build/Definition/ProjectCollection.cs [hunk 5]
```diff
@@ -432,7 +432,7 @@ public static ProjectCollection GlobalProjectCollection
         /// This is the Windows file version (specifically the value of the FileVersion
         /// resource), not necessarily the assembly version.
         /// If you want the assembly version, use Constants.AssemblyVersion.
-        /// This is not the <see cref="ToolsetsVersion">ToolsetCollectionVersion</see>.
+        /// This is not the <see cref="P:Microsoft.Build.BuildEngine.ToolsetCollection.ToolsVersions*">ToolsetCollection.ToolsVersions</see>.
```
Comment (Nirmal4G) [inline src/Build/Definition/ProjectCollection.cs:435]:
```suggestion
        /// This is not the <see cref="ToolsetCollection.ToolsVersions"/>.
```

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 6]
```diff
@@ -53,9 +53,9 @@ public class BuildEventArgsReader : IDisposable
             typeof(BuildEventArgs).GetField("senderName", BindingFlags.Instance | BindingFlags.NonPublic);
 
         /// <summary>
-        /// Initializes a new instance of BuildEventArgsReader using a BinaryReader instance
+        /// Initializes a new instance of <see cref="T:Microsoft.Build.Logging.BuildEventArgsReader"/> using a <see cref="T:System.IO.BinaryReader"/> instance.
```
Comment (Nirmal4G) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:56]:
```suggestion
        /// Initializes a new instance of <see cref="BuildEventArgsReader"/> using a <see cref="BinaryReader"/> instance.
```

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 7]
```diff
@@ -53,9 +53,9 @@ public class BuildEventArgsReader : IDisposable
...
         /// </summary>
-        /// <param name="binaryReader">The BinaryReader to read BuildEventArgs from</param>
+        /// <param name="binaryReader">The <see cref="T:System.IO.BinaryReader"/> to read <see cref="T:Microsoft.Build.Framework.BuildEventArgs"/> from.</param>
```
Comment (Nirmal4G) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:58]:
```suggestion
        /// <param name="binaryReader">The <see cref="BinaryReader"/> to read <see cref="BuildEventArgs"/> from.</param>
```

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 8]
```diff
@@ -79,8 +79,11 @@ public void Dispose()
         internal event Action<BinaryLogRecordKind, byte[]> OnBlobRead;
 
         /// <summary>
-        /// Reads the next log record from the binary reader. If there are no more records, returns null.
+        /// Reads the next log record from the <see cref="T:System.IO.BinaryReader"/>.
```
Comment (Nirmal4G) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:82]:
```suggestion
        /// Reads the next log record from the <see cref="BinaryReader"/>.
```

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 9]
```diff
@@ -79,8 +79,11 @@ public void Dispose()
...
         /// </summary>
+        /// <returns>
+        /// The next <see cref="T:Microsoft.Build.Framework.BuildEventArgs" />. If there are no more records, returns <see langword="null" />.
```
Comment (Nirmal4G) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:85]:
```suggestion
        /// The next <see cref="BuildEventArgs"/>. If there are no more records, returns <see langword="null"/>.
```

File: src/Framework/Sdk/SdkResolver.cs [hunk 10]
```diff
@@ -30,9 +30,12 @@ public abstract class SdkResolver
...
         ///  </returns>   
-        ///  <remarks>
-        ///    Note: You must use <see cref="Microsoft.Build.Framework.SdkResultFactory"/> to return a result.
-        ///  </remarks>
+        ///  <remarks><format type="text/markdown"><![CDATA[
+        ///  ## Remarks
+        ///  > [!NOTE]
+        ///  > You must use the <xref:Microsoft.Build.Framework.SdkResultFactory> to return a result.
+        ///  ]]></format>
+        /// </remarks>
```
Comment (Nirmal4G) [inline src/Framework/Sdk/SdkResolver.cs:38]:
```suggestion
        ///  <remarks>
        ///  Note: You must use the <see cref="SdkResultFactory"/> to return a result.
        /// </remarks>
```

File: src/Framework/Sdk/SdkResolverContext.cs [hunk 11]
```diff
@@ -39,8 +39,11 @@ public abstract class SdkResolverContext
         ///     Version of MSBuild currently running.
         /// </summary>
         /// <remarks>
-        ///     File version based on commit height from our public git repository. This is informational
-        ///     and not equal to the assembly version.
+        ///    <format type="text/markdown"><![CDATA[
+        /// ## Remarks
+        ///     
+        /// File version is informational and not equal to the assembly version.
+        /// ]]></format>
```
Comment (Nirmal4G) [inline src/Framework/Sdk/SdkResolverContext.cs:46]:
```suggestion
        ///     File version is informational and not equal to the assembly version.
```

Do we need to include info about how the file version is calculated?

File: src/Framework/Sdk/SdkResult.cs [hunk 12]
```diff
@@ -9,7 +9,11 @@ namespace Microsoft.Build.Framework
     ///     An abstract interface class to indicate SDK resolver success or failure.
     /// </summary>
     /// <remarks>
-    ///   Note: Use <see cref="Microsoft.Build.Framework.SdkResultFactory"/> to create instances of this class. Do not inherit from this class.
+    ///    <format type="text/markdown"><![CDATA[
+    /// ## Remarks
+    /// > [!NOTE]
+    /// > Use <xref:Microsoft.Build.Framework.SdkResultFactory> to create instances of this class. Do not inherit from this class.
+    /// ]]></format>
```
Comment (Nirmal4G) [inline src/Framework/Sdk/SdkResult.cs:16]:
```suggestion
    ///    Note: Use <see cref="SdkResultFactory"/> to create instances of this class. Do not inherit from this class.
```

File: src/Tasks/Hash.cs [hunk 13]
```diff
@@ -13,8 +13,11 @@ namespace Microsoft.Build.Tasks
     /// Generates a hash of a given ItemGroup items. Metadata is not considered in the hash.
     /// </summary>
     /// <remarks>
-    /// Currently uses SHA1. Implementation subject to change between MSBuild versions. Not
-    /// intended as a cryptographic security measure, only uniqueness between build executions.
+    ///    <format type="text/markdown"><![CDATA[
+    /// ## Remarks
+    /// Currently uses SHA1. The implementation is subject to change between MSBuild versions.
+    /// This class is not intended as a cryptographic security measure, only for uniqueness between build executions.
+    /// ]]></format>
```
Comment (Nirmal4G) [inline src/Tasks/Hash.cs:20]:
```suggestion
    /// Currently uses SHA1. The implementation is subject to change between MSBuild versions.
    /// This class is not intended as a cryptographic security measure, only for uniqueness between build executions.
```

**Review by Nirmal4G — CHANGES_REQUESTED** [review]
Note inside remarks have markdown because of auto conversion but we don't need to copy them exactly as markdown. Just plain text is fine. Same as before.

Also, Type `cref`s are simplified. Some might not have `using`s defined in the file. Check and add necessary `using`s if needed!

**Comment by elachlan** [comment]
> Note inside remarks have markdown because of auto conversion but we don't need to copy them exactly as markdown. Just plain text is fine. Same as before.
> 
> Also, Type `cref`s are simplified. Some might not have `using`s defined in the file. Check and add necessary `using`s if needed!

We have a PR for the cref changes at #7194, do we want to merge that first and then rebase this PR?

**Comment by Nirmal4G** [comment]
@elachlan Then, Can you add these suggestions to that PR? Seems they are not in that PR!?

**Comment by Nirmal4G** [comment]
@elachlan This is merged, so add these suggestions to your `cref` PR.

**Comment by Forgind** [comment]
Do we have somewhere we said we want short crefs? I had no idea, but rainersigwald said no.

**Comment by Forgind** [comment]
If we'd originally listed it as info, that may have been a mistake on our part—if so, sorry about that!

**Comment by rainersigwald** [comment]
I don't feel strongly about the short/long crefs. I do feel strongly that that style issue shouldn't block this change to reduce pain on our docs partners.

**Comment by elachlan** [comment]
@ghogen is the markdown important in this issue? Nirmal's suggestions have been applied in another PR and I wanted to make sure whether or not the markdown was required and needed to be preserved.

**Comment by ghogen** [comment]
@elachlan @Nirmal4G 
The Markdown is used in the dotnet-api-docs repo, but it's only required if there's special formatting such as for a Note/Caution/Tip construct for example. So I think it's OK to implement those suggestions in the other PR.  Note also, we're doing work now to separate the MSBuild API reference from the .NET reference, which will give us more flexibility, because we can control the process better to avoid changes that occur in multiple repos (which is what led to my PR).

**Comment by Nirmal4G** [comment]
@ghogen

> The Markdown is used in the `dotnet-api-docs` repo, but it's only required if there's special formatting such as for a `Note`/`Caution`/`Tip` construct…

The XML Doc comments already includes all these in the `remarks` element but as plain-text with a heading indicating the intent. Wouldn't it be possible for the doc engine to take it and convert to markdown and convert back to plain-text as is? That way, we wouldn't need those `CDATA` elements in the XML Doc.

**Pull request merged**